### PR TITLE
remote-delete: Ask to remove refs

### PR DESCRIPTION
**Pull request by matthiasclasen**
When --force is not given, ask the user for confirmation
whether he wants to remove refs that are installed from
the remote.

Closes: #577

### CODE REVIEW

File: app/flatpak-builtins-remote-delete.c [hunk 1]
```diff
@@ -68,6 +68,33 @@ flatpak_builtin_remote_delete (int argc, char **argv, GCancellable *cancellable,
...
+  if (!opt_force)
+    {
+      g_auto(GStrv) refs = NULL;
+      g_autoptr(GPtrArray) refs_to_remove = NULL;
+      int i;
+
+      refs = flatpak_dir_find_installed_refs (preferred_dir, NULL, NULL, NULL, FLATPAK_KINDS_APP|FLATPAK_KINDS_RUNTIME, 0, NULL);
+      refs_to_remove = g_ptr_array_new_with_free_func (g_free);
+
+      for (i = 0; refs[i]; i++)
```
Comment (mwleeds) [inline app/flatpak-builtins-remote-delete.c]: We have to check if `refs` is NULL before indexing into it right? Might as well also pass an error pointer to `flatpak_dir_find_installed_refs`

File: app/flatpak-builtins-remote-delete.c [hunk 2]
```diff
@@ -68,6 +68,33 @@ flatpak_builtin_remote_delete (int argc, char **argv, GCancellable *cancellable,
...
+        {
+          g_ptr_array_add (refs_to_remove, NULL);
+
+          flatpak_format_choices ((const char **)refs_to_remove->pdata,
+                                  _("The following refs are installed from remote '%s':"), remote_name);
+          if (!flatpak_yes_no_prompt (FALSE, _("Remove them?")))
+            return TRUE;
+        }
+    }
+
```
Comment (mwleeds) [inline app/flatpak-builtins-remote-delete.c]: trailing whitespace

File: app/flatpak-builtins-remote-delete.c [hunk 3]
```diff
@@ -68,6 +68,33 @@ flatpak_builtin_remote_delete (int argc, char **argv, GCancellable *cancellable,
...
+      g_autoptr(GPtrArray) refs_to_remove = NULL;
+      int i;
+
+      refs = flatpak_dir_find_installed_refs (preferred_dir, NULL, NULL, NULL, FLATPAK_KINDS_APP|FLATPAK_KINDS_RUNTIME, 0, NULL);
+      refs_to_remove = g_ptr_array_new_with_free_func (g_free);
+
+      for (i = 0; refs[i]; i++)
+        {
+          g_autofree char *origin = flatpak_dir_get_origin (preferred_dir, refs[i], NULL, NULL);
+          if (strcmp (origin, remote_name) == 0)
```
Comment (mwleeds) [inline app/flatpak-builtins-remote-delete.c]: g_strcmp0

File: app/flatpak-builtins-remote-delete.c [hunk 4]
```diff
@@ -68,6 +68,33 @@ flatpak_builtin_remote_delete (int argc, char **argv, GCancellable *cancellable,
...
+        }
+
+      if (refs_to_remove->len > 0)
+        {
+          g_ptr_array_add (refs_to_remove, NULL);
+
+          flatpak_format_choices ((const char **)refs_to_remove->pdata,
+                                  _("The following refs are installed from remote '%s':"), remote_name);
+          if (!flatpak_yes_no_prompt (FALSE, _("Remove them?")))
+            return TRUE;
```
Comment (mwleeds) [inline app/flatpak-builtins-remote-delete.c]: Don't we need an `else` here to set a variable that will be passed to the second parameter of `flatpak_dir_remove_remote` below (or set `opt_force = TRUE` but that feels a bit wrong)

**Comment by mwleeds** [comment]
Two more potential issues with the code in `flatpak_dir_remove_remote` (not really unique to this PR):
1. There's no call to `flatpak_dir_prune` so you don't reclaim the disk space as you might expect
2. In the documentation for `flatpak_dir_remove_ref`, it says the flatpak must be uninstalled first, but we don't seem to be doing that.

File: app/flatpak-builtins-remote-delete.c [hunk 5]
```diff
@@ -68,7 +69,41 @@ flatpak_builtin_remote_delete (int argc, char **argv, GCancellable *cancellable,
...
```
Comment (mwleeds) [inline app/flatpak-builtins-remote-delete.c]: I wonder if we should return FALSE here with an error message, so the user doesn't think the remote was removed

Comment (matthiasclasen) [reply app/flatpak-builtins-remote-delete.c]: Good point

**Comment by mwleeds** [comment]
> > Two more potential issues with the code in `flatpak_dir_remove_remote` (not really unique to this PR):
> > ```
> > 1. There's no call to `flatpak_dir_prune` so you don't reclaim the disk space as you might expect
> > 
> > 2. In the documentation for `flatpak_dir_remove_ref`, it says the flatpak must be uninstalled first, but we don't seem to be doing that.
> > ```
> 
> What is the effect of this ? Does this mean the installed app stays behind, without a ref backing it up?
> I don't think that was my intention for this interactive approach. I think we want to actually uninstall the apps if the user says yes.
> 
> So this may need some more work.

Yes, I just tested it and the ref gets deleted but the objects in the repo are left behind, and the deploy is left behind, so we need to do a proper uninstall.

File: app/flatpak-builtins-remote-delete.c (hunk 5 above)
Comment (alexlarsson) [reply app/flatpak-builtins-remote-delete.c]: Yeah, that would be good.

**Comment by alexlarsson** [comment]
Yeah, the force opt is kind of "just do it" which may leave things in a weird state. I think in a default setting we should do a proper uninstall transaction first.

**Comment by alexlarsson** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**